seismostats/analysis/declustering/dec_gardner_knopoff.py

```diff
@@ -0,0 +1,153 @@
+        self.fs_time_prop = fs_time_prop
+
+    def _decluster(self, catalog: pd.DataFrame,
+                   ) -> np.ndarray[np.bool_]:
+        """
+        Apply the Gardner-Knopoff declustering algorithm to the catalog.
+
+        The catalog must contain the following columns:
+        - time, magnitude, longitude, latitude
+
+        If there are multiple events with the same magnitude,
+        the earliest event is considered as the mainshock first.
+
+        Args:
+            catalog: the catalog of earthquakes
+
+        Returns:
+            mainshock_flags: boolean array indicating whether the i'th event
+                             is a mainshock
+
+        Raises:
+            ValueError: if a required column is missing
+        """
+        cols = set(("time", "magnitude", "longitude", "latitude"))
+        if not cols.issubset(set(catalog.columns)):
+            raise ValueError("catalog must contain the following columns: "
+                             + ", ".join(cols))
+
+        # each cluster of events is assigned a non-negative integer id
+        cluster_ids = np.zeros(len(catalog), dtype=int)
+        cluster_id = 1
+        magnitude = catalog["magnitude"]
+        longitude = catalog["longitude"]
+        latitude = catalog["latitude"]
+
+        time = catalog["time"].astype("datetime64[s]", errors="ignore").values
+        space_windows, time_windows = self.time_distance_window(magnitude)
+        ordered = catalog[list(cols)].sort_values(by=["magnitude", "time"],
+                                                  ascending=[False, True],
+                                                  kind="mergesort")
+        mainshock_flags = np.ones(len(catalog), dtype=bool)
+        for i, long, lat in zip(ordered.index,
+                                ordered["longitude"],
+                                ordered["latitude"]):
+            # If already assigned to a cluster, skip
+            if cluster_ids[i] != 0:
+                continue
+
+            # Find Events inside both fore- and aftershock time windows
+            dt = time - time[i]
+            vsel = np.logical_and(
+                cluster_ids == 0,
+                np.logical_and(
+                    dt >= (-time_windows[i] * self.fs_time_prop),
+                    dt <= time_windows[i],
+                ),
+            )
+            # Of those events inside time window,
+            # find those inside the distance window
+            vsel1 = (
+                haversine(
+                    longitude[vsel],
+                    latitude[vsel],
+                    long,
+                    lat,
+                )
+                <= space_windows[i]
+            )
+            vsel[vsel] = vsel1
+            # Assign id and flags to this cluster
+            cluster_ids[vsel] = cluster_id
+            cluster_id += 1
+            mainshock_flags[vsel] = 0
+            mainshock_flags[i] = 1
+
+        self.__cluster_ids = cluster_ids
```

Comment:
würde ich entweder in der __init__ der parent oder child class mit `None` instantiaten.